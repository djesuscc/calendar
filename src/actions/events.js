import Swal from 'sweetalert2';
import { fetchToken } from '../helpers/fetch';
import { preparerEvents } from '../helpers/preparerEvents';
import { types } from '../types/types';

export const eventStartNew = (event) => {
  return async (dispatch, getState) => {
    const { uid, name } = getState().auth;
    try {
      const resp = await fetchToken('events', event, 'POST');
      const body = await resp.json();
      if (body.ok) {
        event.id = body.event.id;
        event.user = {
          _id: uid,
          name
        }
        dispatch(eventAddNew(event));
        Swal.fire('Success', 'Event Created', 'success');
      } else {
        Swal.fire('Error', 'Missing Params', 'error');
      }
    } catch (error) {
      console.log(error);
      Swal.fire('Error', 'Please contact admin', 'error');
    }
  }
}

export const eventAddNew = (event) => ({
  type: types.eventAddNew,
  payload: event
});

export const eventSetActive = (event) => ({
  type: types.eventSetActive,
  payload: event
})

export const eventClearActiveEvent = () => ({
  type: types.eventClearActive
})

export const eventStartUpdate = (event) => {
  return async (dispatch) => {
    try {
      const resp = await fetchToken(`events/${event.id}`, event, 'PUT');
      const body = await resp.json();

      if (body.ok) {
        dispatch(eventUpdated(event));
      } else {
        Swal.fire('Error', body.msg, 'error');
      }

    } catch (error) {
      console.log(error);
    }
  }
}

const eventUpdated = (event) => ({
  type: types.eventUpdated,
  payload: event
})

export const eventStartDeleted = () => {
  return async (dispatch, getState) => {
    const { activeEvent } = getState().calendar;
    try {
      const resp = await fetchToken(`events/${activeEvent.id}`, {}, 'DELETE');
      const body = await resp.json();
      if (body.ok) {
        dispatch(eventDeleted());
      } else {
        Swal.fire('Error', body.msg, 'error');
      }
    } catch (error) {
      console.log(error);
    }
  }
}
const eventDeleted = () => ({
  type: types.eventDeleted
})

export const eventStartLoading = () => {
  return async (dispatch) => {
    try {
      const resp = await fetchToken('events');
      const body = await resp.json();
      const events = preparerEvents(body.events);
      dispatch(eventLoaded(events));
    } catch (error) {
      console.log(error);
    }
  }
}

const eventLoaded = (events) => ({
  type: types.eventLoaded,
  payload: events
});

export const eventLogout = () => ({
  type: types.eventLogout
})