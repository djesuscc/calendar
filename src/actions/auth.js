import {
  fetchNoToken,
  fetchToken
} from "../helpers/fetch";
import {
  types
} from "../types/types";
import Swal from 'sweetalert2';
import { eventLogout } from "./events";

export const startLogin = (email, password) => {
  return async (dispatch) => {
    const resp = await fetchNoToken('auth', {
      email,
      password
    }, 'POST');
    const body = await resp.json();

    if (body.ok) {
      localStorage.setItem('token', body.token);
      localStorage.setItem('token-init-date', new Date().getTime());
      dispatch(login({
        uid: body.uid,
        name: body.name
      }));
    } else {
      Swal.fire('Error', body.msg, 'error');
    }
  }
}

export const startRegister = (values) => {
  return async (dispatch) => {
    const resp = await fetchNoToken('auth/register', values, 'POST');
    const {
      ok,
      token,
      msg,
      uid,
      name
    } = await resp.json();

    if (ok) {
      localStorage.setItem('token', token);
      localStorage.setItem('token-init-date', new Date().getTime());
      Swal.fire({
        //title: msg,
        text: 'Register Success',
        icon: 'success',
        showConfirmButton: false,
        timer: 1500
      }).then(() => {
        dispatch(login({
          uid,
          name
        }));
      });
    } else {
      Swal.fire('Error', msg, 'error');
    }
  }
}

export const startChecking = () => {
  return async (dispatch) => {
    const resp = await fetchToken('auth/renew');
    const body = await resp.json();

    if (body.ok) {
      localStorage.setItem('token', body.token);
      localStorage.setItem('token-init-date', new Date().getTime());
      dispatch(login({
        uid: body.uid,
        name: body.name
      }));
    } else {
      dispatch(checkingFinish());
    }

  }
}

const checkingFinish = () => ({
  type: types.authCheckingFinish
})

const login = (user) => ({
  type: types.authLogin,
  payload: user
})

export const startLogout = () => {
  return (dispatch) => {
    localStorage.clear();
    dispatch(logout());  
    dispatch(eventLogout());
  }
}

const logout = () => ({
  type: types.authLogout
})