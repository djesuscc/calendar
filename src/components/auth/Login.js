import React from 'react';
import { useDispatch } from 'react-redux';
import './login.css';
import { useForm } from '../../hooks/useForm';
import { startLogin, startRegister } from '../../actions/auth';
import Swal from 'sweetalert2';

export const Login = () => {
  
  const dispatch = useDispatch();
  const [formLoginValues, handleLoginInputChange] = useForm({
    loginEmail: 'lola@correo.com',
    loginPassword: '123456'
  });

  const [formRegisterValues, handleRegisterInputChange ] = useForm({
    name: '', 
    email: '', 
    password: '',  
    confirm: ''
  })

  const { loginEmail, loginPassword } = formLoginValues;
  const { 
    name, 
    email, 
    password,  
    confirm
  } = formRegisterValues;

  const handleLogin = (e) => {
    e.preventDefault();
    dispatch(startLogin(loginEmail, loginPassword));
  }

  const handleRegister = (e) => {
    e.preventDefault();
    if (password !== confirm) {
      return Swal.fire('Error', 'Passwords are different', 'error');
    }
    dispatch(startRegister(formRegisterValues));
  }

  return (
    <div className="container login-container">
      <div className="row">
        <div className="col-md-6 login-form-1">
          <h3>Login</h3>
          <form onSubmit={handleLogin}>
            <div className="form-group">
              <input
                name="loginEmail"
                type="text"
                className="form-control"
                placeholder="Email"
                value={loginEmail}
                onChange={handleLoginInputChange}
              />
            </div>
            <div className="form-group">
              <input
                name="loginPassword"
                type="password"
                className="form-control"
                placeholder="Password"
                value={loginPassword}
                onChange={handleLoginInputChange}
              />
            </div>
            <div className="form-group">
              <input
                type="submit"
                className="btnSubmit"
                value="Login"
              />
            </div>
          </form>
        </div>

        <div className="col-md-6 login-form-2">
          <h3>Registro</h3>
          <form onSubmit={handleRegister}>
            <div className="form-group">
              <input
                name="name"
                type="text"
                className="form-control"
                placeholder="Name"
                value={name}
                onChange={handleRegisterInputChange}
              />
            </div>
            <div className="form-group">
              <input
                name="email"
                type="email"
                className="form-control"
                placeholder="Email"
                value={email}
                onChange={handleRegisterInputChange}
              />
            </div>
            <div className="form-group">
              <input
                name="password"
                type="password"
                className="form-control"
                placeholder="Password"
                value={password}
                onChange={handleRegisterInputChange}
              />
            </div>

            <div className="form-group">
              <input
                name="confirm"
                type="password"
                className="form-control"
                placeholder="Confirm password"
                value={confirm}
                onChange={handleRegisterInputChange}
              />
            </div>

            <div className="form-group">
              <input
                type="submit"
                className="btnSubmit"
                value="Register" />
            </div>
          </form>
        </div>
      </div>
    </div>
  )
}