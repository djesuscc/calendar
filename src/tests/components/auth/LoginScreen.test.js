import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import Swal from 'sweetalert2';
import '@testing-library/jest-dom';
import { Login } from '../../../components/auth/Login';
import { startLogin, startRegister } from '../../../actions/auth';

jest.mock('../../../actions/auth', () => ({
  startLogin: jest.fn(),
  startRegister: jest.fn()
}));

jest.mock('sweetalert2', () => ({
  fire: jest.fn()
}))

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initState = {};
const store = mockStore(initState);

store.dispatch = jest.fn();
const wrapper = mount(
  <Provider store={ store }>
    <Login />
  </Provider>
)


describe('Test on <LoginScreen />', () => {
  test('should display correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should called login dispatch', () => {
    wrapper.find('input[name="loginEmail"]').simulate('change', {
      target: {
        name: 'loginEmail',
        value: 'lola@correo.com'
      }
    });

    wrapper.find('input[name="loginPassword"]').simulate('change', {
      target: {
        name: 'loginPassword',
        value: '123456'
      }
    });

    wrapper.find('form').at(0).prop('onSubmit')({
      preventDefault(){}
    });

    expect(startLogin).toHaveBeenCalledWith('lola@correo.com', '123456');
  });

  test('There is not record if passwords are different', () => {
    wrapper.find('input[name="name"]').simulate('change', {
      target: {
        name: 'name',
        value: 'Lourdes'
      }
    });

    wrapper.find('input[name="email"]').simulate('change', {
      target: {
        name: 'email',
        value: 'lulu@correo.com'
      }
    });

    wrapper.find('input[name="password"]').simulate('change', {
      target: {
        name: 'password',
        value: '123456'
      }
    });

    wrapper.find('input[name="confirm"]').simulate('change', {
      target: {
        name: 'confirm',
        value: '12345'
      }
    })

    wrapper.find('form').at(1).prop('onSubmit')({
      preventDefault(){}
    });

    expect(startRegister).not.toHaveBeenCalled();
    expect(Swal.fire).toHaveBeenCalledWith('Error', expect.any(String), 'error');
  });
  
  test('Register with same passwords', () => {
    wrapper.find('input[name="name"]').simulate('change', {
      target: {
        name: 'name',
        value: 'Lourdes'
      }
    });

    wrapper.find('input[name="email"]').simulate('change', {
      target: {
        name: 'email',
        value: 'lulu@correo.com'
      }
    });

    wrapper.find('input[name="password"]').simulate('change', {
      target: {
        name: 'password',
        value: '123456'
      }
    });

    wrapper.find('input[name="confirm"]').simulate('change', {
      target: {
        name: 'confirm',
        value: '123456'
      }
    })

    wrapper.find('form').at(1).prop('onSubmit')({
      preventDefault(){}
    });

    expect(startRegister).toHaveBeenCalledWith({name: 'Lourdes', email: 'lulu@correo.com', password: '123456', confirm: '123456'});
    expect(Swal.fire).not.toHaveBeenCalled()
  })
  
  
})