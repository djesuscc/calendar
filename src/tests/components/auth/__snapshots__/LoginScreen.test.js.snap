// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on <LoginScreen /> should display correctly 1`] = `
<div
  className="container login-container"
>
  <div
    className="row"
  >
    <div
      className="col-md-6 login-form-1"
    >
      <h3>
        Login
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="loginEmail"
            onChange={[Function]}
            placeholder="Email"
            type="text"
            value="lola@correo.com"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="loginPassword"
            onChange={[Function]}
            placeholder="Password"
            type="password"
            value="123456"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Login"
          />
        </div>
      </form>
    </div>
    <div
      className="col-md-6 login-form-2"
    >
      <h3>
        Registro
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="name"
            onChange={[Function]}
            placeholder="Name"
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="email"
            onChange={[Function]}
            placeholder="Email"
            type="email"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="password"
            onChange={[Function]}
            placeholder="Password"
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="confirm"
            onChange={[Function]}
            placeholder="Confirm password"
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Register"
          />
        </div>
      </form>
    </div>
  </div>
</div>
`;
